### PR TITLE
feat(Consents): Migrating old consent data to new consents

### PR DESCRIPTION
**Pull request by lauryndbrown**
Fixes Part of https://github.com/carpentries/amy/issues/1881

This migrates the old consent data as is. It is written so that if the consents have been changed the migration does not crash.

### CODE REVIEW

File: amy/consents/migrations/0006_auto_20210614_1234.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
+# Generated by Django 2.2.17 on 2021-06-14 12:34
```
Comment (lauryndbrown) [inline amy/consents/migrations/0006_auto_20210614_1234.py:1]: Note with this migration I'm making assumptions about what is in the database. That the consents and the options I expect to be there are there. However consents can be changed via the admin view, and this is therefore not guarenteed. I don't know what would be a path forward if that is the case.

**Comment by lauryndbrown** [comment]
Testing that it is filled in:

```py
>>> from consents.models import *
>>> from workshops.models import *
>>> person_1 = Person.objects.create(
            personal="Neville",
            family="Longbottom",
            email="person@email.com",
            username="longbottom_neville",
            # # Setting old terms to True to ensure it doesn't affect anything.
            may_contact=True,
            publish_profile=True,
            data_privacy_agreement=True,
            lesson_publication_consent="yes-github"
        )
>>> person_1
<Person: Neville Longbottom <person@email.com>>
>>> [ (c.term.slug, str(c.term_option)) for c in Consent.objects.filter(person=person_1).active().select_related("term", "term_option")]
[('privacy-policy', 'None'), ('may-contact', 'None'), ('public-profile', 'None'), ('may-publish-name', 'None')]
>>> exit()
(env) ➜  amy git:(1881/migrate-old-consent-data-to-new-consents) ✗ python manage.py migrate              Operations to perform:
  Apply all migrations: admin, auth, autoemails, consents, contenttypes, dashboard, django_comments, extrequests, fiscal, reversion, sessions, sites, social_django, workshops
Running migrations:
  Applying consents.0006_auto_20210614_1234... OK
(env) ➜  amy git:(1881/migrate-old-consent-data-to-new-consents) ✗ python manage.py shell  
>>> from workshops.models import *
>>> from consents.models import *
>>> person_1 = Person.objects.get(email="person@email.com")
>>> [ (c.term.slug, str(c.term_option)) for c in Consent.objects.filter(person=person_1).active().select_related("term", "term_option")]
[('privacy-policy', ' (agree)'), ('may-contact', ' (agree)'), ('public-profile', ' (agree)'), ('may-publish-name', 'Yes, and only use my GitHub Handle (agree)')]
```

**Comment by lauryndbrown** [comment]
Ensure they're going to the correct values:

```py
>>> person_2 = Person.objects.create(
            personal="Person",
            family="Two",
            email="person2@email.com",
            username="person2",
            may_contact=False,
            publish_profile=True,
            data_privacy_agreement=False,
            lesson_publication_consent="unset"
        )
>>> [ (c.term.slug, str(c.term_option)) for c in Consent.objects.filter(person=person_1).active().select_related("term", "term_option")]
>>> [ (c.term.slug, str(c.term_option)) for c in Consent.objects.filter(person=person_2).active().select_related("term", "term_option")]
[('privacy-policy', 'None'), ('may-contact', 'None'), ('public-profile', 'None'), ('may-publish-name', 'None')]
>>> exit()
(env) ➜  amy git:(1881/migrate-old-consent-data-to-new-consents) ✗ python manage.py migrate consents 0006
Operations to perform:
  Target specific migration: 0006_auto_20210614_1234, from consents
Running migrations:
  Applying consents.0006_auto_20210614_1234... OK
(env) ➜  amy git:(1881/migrate-old-consent-data-to-new-consents) ✗ python manage.py shell  
>>> from workshops.models import *
>>> from consents.models import *
>>> person_2 = Person.objects.get(email="person2@email.com")
>>> [ (c.term.slug, str(c.term_option)) for c in Consent.objects.filter(person=person_2).active().select_related("term", "term_option")]
[('privacy-policy', 'None'), ('may-contact', ' (decline)'), ('public-profile', ' (agree)'), ('may-publish-name', 'None')]
```

**Comment by lauryndbrown** [comment]
Test that if a term is missing, the migration does not fail:
```
>>> from workshops.models import *
>>> from consents.models import *
>>> person_3 = Person.objects.create(
            personal="Person",
            family="Three",
            email="person3@email.com",
            username="person3",
            may_contact=True,
            publish_profile=True,
            data_privacy_agreement=True,
            lesson_publication_consent="yes-orcid"
        )
>>> [ (c.term.slug, str(c.term_option)) for c in Consent.objects.filter(person=person_3).active().select_related("term", "term_option")]
[('privacy-policy', 'None'), ('may-contact', 'None'), ('public-profile', 'None'), ('may-publish-name', 'None')]
>>> term = Term.objects.get(slug="may-publish-name")
>>> term.archive()
>>> exit()
(env) ➜  amy git:(1881/migrate-old-consent-data-to-new-consents) ✗ python manage.py migrate consents 0006
Operations to perform:
  Target specific migration: 0006_auto_20210614_1234, from consents
Running migrations:
  Applying consents.0006_auto_20210614_1234... OK
(env) ➜  amy git:(1881/migrate-old-consent-data-to-new-consents) ✗ python manage.py shell                
>>> from workshops.models import *
>>> from consents.models import *
>>> person_3 = Person.objects.get(email="person3@email.com")
>>> [ (c.term.slug, str(c.term_option)) for c in Consent.objects.filter(person=person_3).active().select_related("term", "term_option")]
[('privacy-policy', ' (agree)'), ('may-contact', ' (agree)'), ('public-profile', ' (agree)')]
```

**Comment by lauryndbrown** [comment]
If a term option is missing the migration does not fail.

```py
>>> from consents.models import *
>>> from consents.models import *
>>> person_4 = Person.objects.create(
            personal="Person",
            family="Four",
            email="person4@email.com",
            username="person4",
            may_contact=False,
            publish_profile=True,
            data_privacy_agreement=True,
            lesson_publication_consent="yes-orcid"
        )
>>> [ (c.term.slug, str(c.term_option)) for c in Consent.objects.filter(person=person_4).active().select_related("term", "term_option")]
[('privacy-policy', 'None'), ('may-contact', 'None'), ('public-profile', 'None')]
>>> term = Term.objects.get(slug="may-contact")
>>> term.options
<TermOptionQuerySet [<TermOption:  (agree)>, <TermOption:  (decline)>]>
>>> term.options[1]
<TermOption:  (decline)>
>>> term.options[1].archive()
>>> exit()
(env) ➜  amy git:(1881/migrate-old-consent-data-to-new-consents) ✗ python manage.py migrate consents 0006
Operations to perform:
  Target specific migration: 0006_auto_20210614_1234, from consents
Running migrations:
  Applying consents.0006_auto_20210614_1234... OK
(env) ➜  amy git:(1881/migrate-old-consent-data-to-new-consents) ✗ python manage.py shell 
>>> from workshops.models import *
>>> from consents.models import *
>>> person_4 = Person.objects.get(email="person4@email.com")
>>> [ (c.term.slug, str(c.term_option)) for c in Consent.objects.filter(person=person_4).active().select_related("term", "term_option")]
[('privacy-policy', ' (agree)'), ('may-contact', 'None'), ('public-profile', ' (agree)')]
```

**Comment by lauryndbrown** [comment]
note for @pbanaszkiewicz  my linter doesn't appear to run on the file.

**Comment by pbanaszkiewicz** [comment]
> 
> 
> note for @pbanaszkiewicz my linter doesn't appear to run on the file during pre-commit. Is it still working?
> 
> Oh I see. Maybe it skips migrations locally but not when github actions is run
> 
> ```
> black................................................(no files to check)Skipped
> flake8...............................................(no files to check)Skipped
> isort....................................................................Passed
> [1881/migrate-old-consent-data-to-new-consents e1847b95] Changed lint issues
>  1 file changed, 9 insertions(+), 9 deletions(-)
> ```

Thanks for this report. I created this issue: https://github.com/carpentries/amy/issues/1987

File: amy/consents/migrations/0006_auto_20210614_1234.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+def copy_may_contact(apps, schema_editor):
+    Consent = apps.get_model("consents", "Consent")
+    TermOption = apps.get_model("consents", "TermOption")
+
+    term_slug = "may-contact"
+    try:
+        options = TermOption.objects.filter(term__slug=term_slug, archived_at=None)
+        may_contact_agree = [
+            option for option in options if option.option_type == AGREE
+        ][0]
```
Comment (pbanaszkiewicz) [inline amy/consents/migrations/0006_auto_20210614_1234.py:50]: I've commented on such construct before. This calls for `next()`.

File: amy/consents/migrations/0006_auto_20210614_1234.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+    for old_consent in Consent.objects.filter(
+        term__slug=term_slug, archived_at=None
+    ).select_related("person"):
+        if old_consent.person.may_contact:
+            reconsent(Consent, old_consent, may_contact_agree)
+        else:
+            reconsent(Consent, old_consent, may_contact_disagree)
+
+
+def copy_public_profile(apps, schema_editor):
```
Comment (pbanaszkiewicz) [inline amy/consents/migrations/0006_auto_20210614_1234.py:66]: The functions are very similar to each other. Perhaps in future you could try to use `partial` to reduce amount of code.

**Review by pbanaszkiewicz — APPROVED** [review]
I left some minor comments to consider in future. Looks good to me.

File: amy/consents/migrations/0006_auto_20210614_1234.py (hunk 3 above)
Comment (lauryndbrown) [reply amy/consents/migrations/0006_auto_20210614_1234.py:66]: That makes sense. Thanks for the tip.

File: amy/consents/migrations/0006_auto_20210614_1234.py (hunk 2 above)
Comment (lauryndbrown) [reply amy/consents/migrations/0006_auto_20210614_1234.py:50]: @pbanaszkiewicz I may have missed your comment. Where would next have been used here?

**Pull request merged**